src/python/grpcio/grpc/experimental/aio/_call.py

```diff
@@ -308,16 +308,17 @@ def cancel(self) -> bool:
                                 _LOCAL_CANCELLATION_DETAILS, None, None))
 
     def __await__(self) -> ResponseType:
-        """Wait till the ongoing RPC request finishes.
-
-        Returns:
-          Response of the RPC call.
-
-        Raises:
-          RpcError: Indicating that the RPC terminated with non-OK status.
-          asyncio.CancelledError: Indicating that the RPC was canceled.
-        """
-        response = yield from self._call
+        """Wait till the ongoing RPC request finishes."""
+        try:
+            response = yield from self._call
+        except asyncio.CancelledError:
+            # Even if we converted all other CancelledError, there is still
+            # this corner case. If the application cancels immediately after
+            # the Call object is created, we will observe this
+            # `CancelledError`.
+            if not self.cancelled():
+                self.cancel()
+            raise _create_rpc_error(_EMPTY_METADATA, self._status.result())
```

Comment:
With that, we would be masking `asyncio.CancelledError`,  which IMO could be nocive for the users. So, not allowing them to react upon an `asyncio.CancelledError` exception or even worst, having them reacting to a `RpcError` which was originally caused because by an `asyncio.CancelledError` exception.

My opinion here is that graceful cancelations - the ones explicitly asked by the caller using the `call.cancel()` interface - would need to end up as simple `RpcErrors`, while implicit cancellations because of an `asyncio.CancelledError` would not be treated as RPC errors and do our best.

Do our best could mean either of the following things:

1- Cancel the ongoing RPC here [1] and do nothing else, the call object would end up in a none finished state.
2 - The same as 1 but reacting upon a `asyncio.CancelledError` for leaving the call object in a finished state but reraising always the original exception.

Theoretically from the point of view of the caller, if there are no cleanup tasks to do they should not take care of `asyncio.CancelledError` exceptions, so having something like this:

```python
try:
   await call
except grpc.RpcError as err:
   assert await call.code() == err.code()
   if err.code() == CANCELLED:
       # should I retry?
``` 

[1] https://github.com/grpc/grpc/pull/21506/files#diff-d07d1d5928e4bc4defeb2e1218c57186R154

Reply:
I agree with you that this solution does have a drawback that it disabled users to react upon `asyncio.CancelledError`. Also, it affects the logging, since `CancelledError` might be silent in certain cases but `RpcError` never get silent automatically.

I understand your proposal, but there is a topic of predicability. If we allow `asyncio.CancelledError` to be raised from our package, then users need to always remember handling both the `asyncio.CancelledError` and the `RpcError`. For RPC in different state, the outcome exception will be different. I think it is not an ideal design.

As discussed before, no matter the cancellation is originated locally or remotely, the retry logic always require spawning a retry RPC. So, by unifying the exceptions, we make it less error-prone (one less try-catch clause).

```Python
try:
   await call
except grpc.RpcError as err:
   if err.code() == CANCELLED:
       # apply retry logic
```

Reply:
Not convinced :) see my last comment here 
https://github.com/grpc/grpc/pull/21506/files#r360149977

Reply:
I agree with @pfreixes on this one. There's value in being able to distinguish between the two.